unified-runtime/source/adapters/opencl/adapter.cpp

```diff
@@ -42,45 +47,38 @@ ur_adapter_handle_t_::ur_adapter_handle_t_() {
 #undef CL_CORE_FUNCTION
 
 #endif // _MSC_VER
+  assert(!liveAdapter);
+  liveAdapter = this;
 }
 
-static ur_adapter_handle_t adapter = nullptr;
+ur_adapter_handle_t_::~ur_adapter_handle_t_() {
+  assert(liveAdapter == this);
+  liveAdapter = nullptr;
+}
 
 ur_adapter_handle_t ur::cl::getAdapter() {
-  if (!adapter) {
+  if (!liveAdapter) {
     die("OpenCL adapter used before initalization or after destruction");
   }
-  return adapter;
-}
-
-static void globalAdapterShutdown() {
-  if (cl_ext::ExtFuncPtrCache) {
-    delete cl_ext::ExtFuncPtrCache;
-    cl_ext::ExtFuncPtrCache = nullptr;
-  }
-  if (adapter) {
-    delete adapter;
-    adapter = nullptr;
-  }
+  return liveAdapter;
 }
 
 UR_APIEXPORT ur_result_t UR_APICALL
 urAdapterGet(uint32_t NumEntries, ur_adapter_handle_t *phAdapters,
              uint32_t *pNumAdapters) {
+  static std::mutex AdapterConstructionMutex{};
+
   if (NumEntries > 0 && phAdapters) {
-    // Sometimes urAdaterGet may be called after the library already been torn
-    // down, we also need to create a temporary handle for it.
-    if (!adapter) {
-      adapter = new ur_adapter_handle_t_();
-      atexit(globalAdapterShutdown);
-    }
+    std::lock_guard<std::mutex> Lock{AdapterConstructionMutex};
 
-    std::lock_guard<std::mutex> Lock{adapter->Mutex};
-    if (adapter->RefCount++ == 0) {
-      cl_ext::ExtFuncPtrCache = new cl_ext::ExtFuncPtrCacheT();
+    if (!liveAdapter) {
+      *phAdapters = new ur_adapter_handle_t_();
+    } else {
+      *phAdapters = liveAdapter;
     }
 
-    *phAdapters = adapter;
+    auto &adapter = *phAdapters;
+    adapter->RefCount++;
```

Comment:
there was a similar pattern to this in cuda/hip and I changed it so the declaration of RefCount in the header is `= 1` instead of `= 0`, seems cleaner

Reply:
oh yeah it never occurred to me that that AdapterGet is a refcount increment, too used to thinking about anything above context as returning a pointer and nothing else

Reply:
actually should it be a refcount increment? not a question for friday evening but a question worth asking

Reply:
What do you mean? what else could it be?

Reply:
Because of the name I always thought of it like a pure `Get`, as in it just returns the handle without changing the ref count (other than initializing it to one on the first call). The spec does say it's a refcount increment though so never mind
> An application may call this entry point multiple times to acquire multiple references to the adapter handle(s).